onnxruntime/core/providers/cpu/tensor/scatter.cc

```diff
@@ -0,0 +1,188 @@
+
+class Scatter final : public OpKernel {
+ public:
+  Scatter(const OpKernelInfo& info) : OpKernel(info) {
+    ORT_ENFORCE(info.GetAttr<int64_t>("axis", &axis_).IsOK(),
+                "Missing/Invalid 'axis' attribute value");
+  }
+  ~Scatter() = default;
+  Status Compute(OpKernelContext* context) const override;
+
+ private:
+  int64_t axis_;
+};
+
+ONNX_CPU_OPERATOR_KERNEL(
+    Scatter,
+    9,
+    KernelDefBuilder()
+        .MayInplace(0, 0)
+        .TypeConstraint("T", DataTypeImpl::AllTensorTypes())
+        .TypeConstraint("Tind", std::vector<MLDataType>{DataTypeImpl::GetTensorType<int32_t>(), DataTypeImpl::GetTensorType<int64_t>()}),
+    Scatter);
+
+template <class Tin>
+Status CopyScatterData(const Tensor* data_input, const Tensor* indices_input, const Tensor* updates_input,
+                       const int64_t axis, Tensor* data_output) {
+  const TensorShape& input_data_shape = data_input->Shape();
+  const Tin* indices_data = indices_input->template Data<Tin>();
+  const auto num_indices = indices_input->Shape().Size();
+  for (int64_t i = 0; i < num_indices; ++i) {
+    Tin idx = indices_data[i];
+    if (idx < 0 || idx >= input_data_shape[axis]) {
+      return ORT_MAKE_STATUS(ONNXRUNTIME, INVALID_ARGUMENT, "indices element out of data bounds, idx=", idx,
+                             " data_dim=", input_data_shape[axis]);
+    }
+  }
+
+  const auto input_elements = input_data_shape.Size();
+  const auto element_bytes = data_input->DataType()->Size();
+  const auto total_input_bytes = data_input->Size();
+
+  const uint8_t* src_base = reinterpret_cast<const uint8_t*>(data_input->DataRaw());
+  uint8_t* dst_base = reinterpret_cast<uint8_t*>(data_output->MutableDataRaw());
+  const bool is_string_type = data_input->DataType() == DataTypeImpl::GetType<std::string>();
+
+  // We allow runtime to re-use input for output. If input/output Tensor* are the same
+  // we do not copy
+  if (src_base != dst_base) {
+    if (is_string_type) {
+      const std::string* str_begin = data_input->template Data<std::string>();
+      const std::string* str_end = str_begin + input_elements;
+      std::string* dst = data_output->template MutableData<std::string>();
+      std::copy(str_begin, str_end, dst);
+    } else {
+      memcpy(dst_base, src_base, total_input_bytes);
+    }
+  }
+
+  // Now poke updates
+
+  const auto& upd_shape = updates_input->Shape();
+  const auto num_dims = input_data_shape.NumDimensions();
+  assert(num_dims > 0);
+
+  // Allocate and zero out counts. We are expecting that input and
+  // update ranks are the same so the vectors of equal length
+  // This is validated
+  std::vector<int64_t> dim_counters(num_dims);
+
+  // Create weights for each input dimension so we can map
+  // update dimensions to input/output dimensions
+  std::vector<int64_t> out_weights(num_dims);
+
+  out_weights.back() = 1;
+  if (out_weights.size() > 1) {
+    for (int64_t i = int64_t(num_dims - 2); i >= 0; --i) {
```

Comment:
>num_dims - 2 [](start = 29, length = 12)

A comment explaining why 'num_dims -2' is used would help someone new to the code.